# Detect Phishing URLs
### Capstone 3 - Data Wrangling and EDA
Michael Garber

#### High-Level Steps
1. Data Wrangling
    1. Data Collection
    2. Data Organization
    3. Data Definition
    4. Data Cleaning
1. Exploratory Data Analysis


#### Data Wrangling

##### Data Collection
URL Data
- https://www.kaggle.com/datasets/hassaanmustafavi/phishing-urls-dataset

TLD (Top-level Domain) Data
- https://www.iana.org/domains/root/db

In [6]:
# Please Install KaggleHub if you don't already have it
%pip install kagglehub

Note: you may need to restart the kernel to use updated packages.


In [7]:
# Import Packages
import pandas as pd
import kagglehub
import os
import shutil
from urllib.parse import urlparse

In [8]:
# Download Data
kPath = kagglehub.dataset_download("hassaanmustafavi/phishing-urls-dataset")

In [9]:
# build source file path
fileName = 'url_dataset.csv'
srcFilePath = os.path.join(kPath, fileName)
dataDir = os.path.join(os.pardir, 'data', 'raw', fileName)

# Move File to project folder
shutil.copy(srcFilePath, dataDir)

'..\\data\\raw\\url_dataset.csv'

In [10]:
# Create pandas dataframe - urlData
urlData = pd.read_csv(dataDir)

In [11]:
# Create pandas dataframe - TLDs (Top-level Domains)
tldDir = '../data/raw/TLDs.csv'
tldNames = pd.read_csv(tldDir)

# TLDs source...
# https://www.iana.org/domains/root/db

In [12]:
# check dataframe urlData - # or rows
print(urlData.shape)

(450176, 2)


In [13]:
# check dataframe tldNames - # or rows
print(tldNames.shape)

(1591, 3)


##### Data Organization
Project file structure based on the cookiecutter data science template. \
[https://drivendata.github.io/cookiecutter-data-science/](https://drivendata.github.io/cookiecutter-data-science/)

Folder structure tree (GitHub) \
[https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md](https://github.com/mdgarber/DetectPhishURL/blob/7dd7d38c001590b4629f8810906a3724ab107fd5/DetectPhishURL/README.md)

##### Data Definition

- Column names
- Data types
- Description of the columns
- Counts and unique values
- Ranges of values
- Calc Summary statistics

In [16]:
# Data types, unique values, range of index
urlData.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450176 entries, 0 to 450175
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   url     450176 non-null  object
 1   type    450176 non-null  object
dtypes: object(2)
memory usage: 6.9+ MB


Description of the columns
- url - The web address to be analyzed.
- type - The classification of the URL (phishing or legitimate).

In [18]:
# increase pandas dataframe column width - full size
pd.set_option('display.width', None)

In [19]:
# Value counts for URLs
pd.DataFrame(urlData['url' ].value_counts())

count
url                                                      
https://www.google.com                                  1
https://www.tabheaven.com/gales-eric-tabs.html          1
https://www.tabs-database.com/justin-mcroberts-...      1
https://www.tabpower.com/a806.html                      1
https://www.tabor.edu/alumni/directory?decade=1960      1
...                                                   ...
https://www.billboard.com/artist/anita-pointer/...      1
https://www.billboard.com/artist/andy-kim/chart...      1
https://www.billboard.com/artist/alvino-rey/dis...      1
https://www.billboard.com/artist/3x-krazy/disco...      1
http://writeassociate.com/test/Portal/inicio/IO...      1

[450176 rows x 1 columns]

- These URLs above are all unique ^

In [21]:
pd.DataFrame(urlData['type' ].value_counts()).head()

count
type              
legitimate  345738
phishing    104438

In [22]:
# describe dataframe
urlData.describe()

url        type
count                   450176      450176
unique                  450176           2
top     https://www.google.com  legitimate
freq                         1      345738

Data Definition Summary

- Column names
    - **url**
        - Data types
            - series of string
        - Description of the column 
            - The web address to be analyzed.
        - Counts and unique values 
            - 450176 (all unique)
        - Ranges of values 
            - all unique URLs
    - **type**
        - Data types 
            - series of string
        - Description of the column
            - The classification of the URL (phishing or legitimate).
        - Counts and unique values
            - legitimate    345738
            - phishing      104438
        - Ranges of values
            - phishing
            - legitimate
            
- Calc Summary statistics
    - N/A ... text data

##### Data Cleaning

- this dataset is mostly clean  =)
- identify invalid URLs (those that urllib couldn't parse) and remove

#### Exploratory Data Analysis

- create new features to parse URLs
- do value counts on the new url parts
- check correlations of the url parts to the target variable

In [26]:
# view the data head - URLs
urlData.head(10)

url        type
0     https://www.google.com  legitimate
1    https://www.youtube.com  legitimate
2   https://www.facebook.com  legitimate
3      https://www.baidu.com  legitimate
4  https://www.wikipedia.org  legitimate
5     https://www.reddit.com  legitimate
6      https://www.yahoo.com  legitimate
7   https://www.google.co.in  legitimate
8         https://www.qq.com  legitimate
9     https://www.amazon.com  legitimate

In [27]:
# view the data head - only phish
urlData[urlData['type'] == 'phishing'].head(10)

url      type
345738                   http://atualizacaodedados.online  phishing
345739                     http://webmasteradmin.ukit.me/  phishing
345740           http://stcdxmt.bigperl.in/klxtv/apps/uk/  phishing
345741   https://tubuh-syarikat.com/plugins/fields/files/  phishing
345742  http://rolyborgesmd.com/exceword/excel.php?.ra...  phishing
345743  http://ongelezen-voda.000webhostapp.com/inlogg...  phishing
345744  http://www.valenzaceramic.com/home/webapps/e52...  phishing
345745  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345746  http://membership-issue.forteimpex.com/dk2mmm=...  phishing
345747  http://chronopost-service-enligne.net/56123s/r...  phishing

##### TLD Data
- to be integrated with url data set

In [29]:
# view the data (to be used as metadata for URLs)
tldNames

Domain          Type  \
0         .aaa       generic   
1        .aarp       generic   
2      .abarth       generic   
3         .abb       generic   
4      .abbott       generic   
...        ...           ...   
1586    .zippo       generic   
1587       .zm  country-code   
1588     .zone       generic   
1589  .zuerich       generic   
1590       .zw  country-code   

                                            TLD Manager  
0                 American Automobile Association, Inc.  
1                                                  AARP  
2                                          Not assigned  
3                                               ABB Ltd  
4                             Abbott Laboratories, Inc.  
...                                                 ...  
1586                                       Not assigned  
1587  Zambia Information and Communications Technolo...  
1588                                    Binky Moon, LLC  
1589                   Kanton Zürich (Canton of Zurich)  
1590  Postal and Telecommunications Regulatory Autho...  

[1591 rows x 3 columns]

##### URL Parsing

In [31]:
# create function urlparse_try 
#    otherwise, parsing will fail on an invalid URL
def urlparse_try(url):
    try:
        return urlparse(url)
    except:
        return 'invalid URL'

In [32]:
# engineer new feature for EDA - parsedUrl
urlData['parsedUrl'] = urlData['url'].apply(urlparse_try)

In [33]:
# example of a parsed url
print('Example of a parsed url \n')
print(urlparse(urlData['url'][1]))

Example of a parsed url 

ParseResult(scheme='https', netloc='www.youtube.com', path='', params='', query='', fragment='')


Parsed URL parts
> scheme://netloc/path;parameters?query#fragment

> **URL Part Reference**

![Image](../references/URL_part_diagram.png)

- image from https://www.geeksforgeeks.org/components-of-a-url/


In [36]:
# find invalid URLs (urls that failed urllib parsing...those will cause errors)
invalidUrlRows = urlData[urlData['parsedUrl'] == 'invalid URL']
pd.DataFrame(invalidUrlRows).head()

url      type  \
397556  http://ladiesfirst-privileges[.]com/656465/d56...  phishing   

          parsedUrl  
397556  invalid URL

In [37]:
# remove invalid URLs (they will causes errors)
urlData = urlData.drop(invalidUrlRows.index)

In [38]:
# confirm 1 record removed
urlData.shape

(450175, 3)

In [39]:
# engineer new features from parsedUrl

# scheme
urlData['urlPart_scheme'] = urlData['parsedUrl'].apply(lambda urlObj : urlObj.scheme)
#subdomain
#domain
#top level domain
#port
#path
#query
#fragement

In [40]:
urlData.head()

url        type                           parsedUrl  \
0     https://www.google.com  legitimate     (https, www.google.com, , , , )   
1    https://www.youtube.com  legitimate    (https, www.youtube.com, , , , )   
2   https://www.facebook.com  legitimate   (https, www.facebook.com, , , , )   
3      https://www.baidu.com  legitimate      (https, www.baidu.com, , , , )   
4  https://www.wikipedia.org  legitimate  (https, www.wikipedia.org, , , , )   

  urlPart_scheme  
0          https  
1          https  
2          https  
3          https  
4          https

In [41]:
# split top 30 hostnames example
urlData['parsedUrl'].head(30).apply(lambda domainName : domainName[1].split(sep='.'))

0          [www, google, com]
1         [www, youtube, com]
2        [www, facebook, com]
3           [www, baidu, com]
4       [www, wikipedia, org]
5          [www, reddit, com]
6           [www, yahoo, com]
7       [www, google, co, in]
8              [www, qq, com]
9          [www, amazon, com]
10         [www, taobao, com]
11        [www, twitter, com]
12          [www, tmall, com]
13      [www, google, co, jp]
14             [www, vk, com]
15           [www, live, com]
16      [www, instagram, com]
17           [www, sohu, com]
18       [www, sina, com, cn]
19             [www, jd, com]
20          [www, weibo, com]
21             [www, 360, cn]
22          [www, google, de]
23      [www, google, co, uk]
24     [www, google, com, br]
25          [www, google, fr]
26          [www, google, ru]
27        [www, netflix, com]
28          [www, yandex, ru]
29    [www, list, tmall, com]
Name: parsedUrl, dtype: object

In [206]:
# create variable for the split u[ url domain names
hostNameSplit = pd.DataFrame(urlData['parsedUrl'].apply(lambda domainName : domainName[1].split(sep='.')))

In [207]:
# count the number of URL parts - (e.x. www.google.co.uk  is *4* URL parts)
hostNameSplit['numUrlParts'] = hostNameSplit['parsedUrl'].apply(lambda domainSplit : len(domainSplit))

In [238]:
# view
hostNameSplit.head()

parsedUrl  numUrlParts
0     [www, google, com]            3
1    [www, youtube, com]            3
2   [www, facebook, com]            3
3      [www, baidu, com]            3
4  [www, wikipedia, org]            3

In [244]:
# explore the variation in hostname parts - not equal to the standard 3
hostNameSplit[hostNameSplit['numUrlParts'] != 3]

parsedUrl  numUrlParts
7        [www, google, co, in]            4
13       [www, google, co, jp]            4
18        [www, sina, com, cn]            4
23       [www, google, co, uk]            4
24      [www, google, com, br]            4
...                        ...          ...
450171          [ecct-it, com]            2
450172        [faboleena, com]            2
450173        [faboleena, com]            2
450174       [atualizapj, com]            2
450175   [writeassociate, com]            2

[157402 rows x 2 columns]

##### Correlation to Target

In [46]:
# relationship between http vs https (secure) and type (phish or not)

In [47]:
# relationship between top level domain and type (phish or not)
# 

In [48]:
# Export df to file
urlData.to_csv('../data/interim/urlData_raw.csv')

- the dataset is very simple and not well suited for statistical analysis in it's current form
- will parse the URLs and perform feature engineering in the next notebook for pre-processing

- play with the data a bit more for EDA
- look at relationship between target and
    - https and http urls
    - .com and .com.in (country specific domains)
    - any other initial analysis that can be done without extensive feature engineering (that will be done in next notebook)
    - resources
        - https://www.geeksforgeeks.org/components-of-a-url/
        - https://www.techtarget.com/whatis/definition/named-entity-recognition-NER   (from Kevin)

In [50]:
'''
TO DO
-split out url parts from hostname - domain, top level domain, sub-level domain
-join the TLD columns
    ex....
    merged_left = pd.merge(df_left, df_right, on='key', how='left')
    print("\nLeft Join (on 'key'):\n", merged_left)
-Use hostname to check if url part is IP address or hostname
-need to write function to correctly split and *assign* hostname parts to either subdomain, domain, or top level domain based on number of periods (or parts). maybe use a case statement
-do additional EDA
'''

'\nTO DO\n-split out url parts from hostname - domain, top level domain, sub-level domain\n-join the TLD columns\n    ex....\n    merged_left = pd.merge(df_left, df_right, on=\'key\', how=\'left\')\n    print("\nLeft Join (on \'key\'):\n", merged_left)\n-Use hostname to check if url part is IP address or hostname\n-need to write function to correctly split and *assign* hostname parts to either subdomain, domain, or top level domain based on number of periods (or parts). maybe use a case statement\n-do additional EDA\n'